File: bench/arena/bench_arena.go
// bench_arena.go — Go 综合竞技场基准
//
// 五个赛道：HashMap / Sort / String / JSON / Pool
// 输出 benchstat 兼容格式
package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"sort"
	"strconv"
	"strings"
	"sync"
	"testing"
)

const (
	HASHMAP_N     = 100000
	HASHMAP_ITERS = 100
	SORT_N        = 1000
	SORT_ITERS    = 100
	STRING_N      = 10000
	STRING_ITERS  = 100
	JSON_N        = 1000
	JSON_ITERS    = 100
)

// === 赛道 1: HashMap ===

func BenchmarkHashMap_Insert(b *testing.B) {
	keys := make([]int, HASHMAP_N)
	for i := range keys {
		keys[i] = (i * 2654435761) & 0x7FFFFFFF
	}
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		m := make(map[int]int, HASHMAP_N)
		for j := 0; j < HASHMAP_N; j++ {
			m[keys[j]] = j
		}
	}
}

func BenchmarkHashMap_Lookup(b *testing.B) {
	keys := make([]int, HASHMAP_N)
	m := make(map[int]int, HASHMAP_N)
	for i := range keys {
		keys[i] = (i * 2654435761) & 0x7FFFFFFF
		m[keys[i]] = i
	}
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		found := 0
		for j := 0; j < HASHMAP_N; j++ {
			if _, ok := m[keys[j]]; ok {
				found++
			}
		}
	}
}

func BenchmarkHashMap_Iterate(b *testing.B) {
	m := make(map[int]int, HASHMAP_N)
	for i := 0; i < HASHMAP_N; i++ {
		m[(i*2654435761)&0x7FFFFFFF] = i
	}
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		count := 0
		for range m {
			count++
		}
	}
}

// === 赛道 2: Sort ===

func BenchmarkSort_Int32(b *testing.B) {
	data := make([]int, SORT_N)
	for i := range data {
		data[i] = (i * 48271) % 1000000
	}
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		d := make([]int, SORT_N)
		copy(d, data)
		sort.Ints(d)
	}
}

// === 赛道 3: String ===

func BenchmarkString_Builder(b *testing.B) {
	for n := 0; n < b.N; n++ {
		var sb strings.Builder
		sb.Grow(STRING_N * 16)
		for i := 0; i < STRING_N; i++ {
			sb.WriteString("item_")
			sb.WriteString(strconv.Itoa(i))
			sb.WriteByte(',')
		}
		_ = sb.String()
	}
}

func BenchmarkString_Concat(b *testing.B) {
	for n := 0; n < b.N; n++ {
		result := ""
		for i := 0; i < STRING_N; i++ {
			result = result + "item_" + strconv.Itoa(i) + ","
		}
	}
}

// === 赛道 4: JSON ===

type User struct {
	ID    int    `json:"id"`
	Name  string `json:"name"`
	Email string `json:"email"`
	Age   int    `json:"age"`
}

type Users struct {
	Users []User `json:"users"`
}

func BenchmarkJSON_Parse(b *testing.B) {
	users := Users{Users: make([]User, JSON_N)}
	for i := 0; i < JSON_N; i++ {
		users.Users[i] = User{
			ID:    i,
			Name:  "user_" + strconv.Itoa(i),
			Email: "user" + strconv.Itoa(i) + "@example.com",
			Age:   20 + (i % 50),
		}
	}
	data, _ := json.Marshal(users)

	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		var u Users
		json.Unmarshal(data, &u)
	}
}

// === 赛道 5: Pool (sync.Pool) ===

type PoolItem struct {
	Value int
}

func BenchmarkPool_SingleThread(b *testing.B) {
	pool := sync.Pool{
		New: func() any { return &PoolItem{} },
	}
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		for i := 0; i < 10000; i++ {
			item := pool.Get().(*PoolItem)
			item.Value = i
			pool.Put(item)
		}
	}
}

func BenchmarkPool_Concurrent_8(b *testing.B) {
	pool := sync.Pool{
		New: func() any { return &PoolItem{} },
	}
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		var wg sync.WaitGroup
		for t := 0; t < 8; t++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				for i := 0; i < 100000; i++ {
					item := pool.Get().(*PoolItem)
					item.Value = i
					pool.Put(item)
				}
			}()
		}
		wg.Wait()
	}
}

func main() {
	fmt.Println("=== Go Arena Benchmark ===")
	fmt.Println("Run with: go test -bench=. -benchmem -count=6 | tee go_arena.txt")
	fmt.Println("Then: benchstat go_arena.txt")
}

File: bench/arena/bench_arena.rs
// bench_arena.rs — Rust 综合竞技场基准
//
// 七个赛道：HashMap / Sort / String / JSON / TOML / Regex
// 使用 criterion 运行: cargo bench
use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion, Throughput};
use std::collections::HashMap;
use std::fmt::Write;

const HASHMAP_N: usize = 100000;
const SORT_N: usize = 10000;
const STRING_N: usize = 10000;
const JSON_N: usize = 1000;
const TOML_N: usize = 500;
const REGEX_N: usize = 1000;

// === 赛道 1: HashMap ===

fn bench_hashmap_insert(c: &mut Criterion) {
    let mut group = c.benchmark_group("HashMap");
    group.throughput(Throughput::Elements(HASHMAP_N as u64));

    group.bench_function("Insert", |b| {
        b.iter(|| {
            let mut m = HashMap::with_capacity(HASHMAP_N);
            for i in 0..HASHMAP_N {
                m.insert((i as u64 * 2654435761) & 0x7FFFFFFF, i);
            }
            m
        })
    });

    group.finish();
}

fn bench_hashmap_lookup(c: &mut Criterion) {
    let mut m = HashMap::with_capacity(HASHMAP_N);
    for i in 0..HASHMAP_N {
        m.insert((i as u64 * 2654435761) & 0x7FFFFFFF, i);
    }
    let keys: Vec<u64> = (0..HASHMAP_N)
        .map(|i| (i as u64 * 2654435761) & 0x7FFFFFFF)
        .collect();

    let mut group = c.benchmark_group("HashMap");
    group.throughput(Throughput::Elements(HASHMAP_N as u64));

    group.bench_function("Lookup", |b| {
        b.iter(|| {
            let mut found = 0usize;
            for key in &keys {
                if m.contains_key(key) {
                    found += 1;
                }
            }
            found
        })
    });

    group.finish();
}

fn bench_hashmap_iterate(c: &mut Criterion) {
    let mut m = HashMap::with_capacity(HASHMAP_N);
    for i in 0..HASHMAP_N {
        m.insert((i as u64 * 2654435761) & 0x7FFFFFFF, i);
    }

    let mut group = c.benchmark_group("HashMap");
    group.throughput(Throughput::Elements(HASHMAP_N as u64));

    group.bench_function("Iterate", |b| {
        b.iter(|| {
            let mut count = 0usize;
            for _ in &m {
                count += 1;
            }
            count
        })
    });

    group.finish();
}

// === 赛道 2: Sort ===

fn bench_sort_int32(c: &mut Criterion) {
    let data: Vec<i32> = (0..SORT_N).map(|i| ((i as i64 * 48271) % 1000000) as i32).collect();

    let mut group = c.benchmark_group("Sort");
    group.throughput(Throughput::Bytes((SORT_N * 4) as u64));

    group.bench_function("Int32", |b| {
        b.iter(|| {
            let mut d = data.clone();
            d.sort_unstable();
            d
        })
    });

    group.finish();
}

// === 赛道 3: String ===

fn bench_string_builder(c: &mut Criterion) {
    let mut group = c.benchmark_group("String");
    group.throughput(Throughput::Bytes((STRING_N * 16) as u64));

    group.bench_function("Builder", |b| {
        b.iter(|| {
            let mut s = String::with_capacity(STRING_N * 16);
            for i in 0..STRING_N {
                s.push_str("item_");
                s.push_str(&i.to_string());
                s.push(',');
            }
            s
        })
    });

    group.finish();
}

fn bench_string_concat(c: &mut Criterion) {
    let mut group = c.benchmark_group("String");
    group.throughput(Throughput::Bytes((STRING_N * 16) as u64));

    group.bench_function("Concat", |b| {
        b.iter(|| {
            let mut result = String::new();
            for i in 0..STRING_N {
                result = result + "item_" + &i.to_string() + ",";
            }
            result
        })
    });

    group.finish();
}

// === 赛道 4: JSON ===

#[derive(serde::Serialize, serde::Deserialize)]
struct User {
    id: usize,
    name: String,
    email: String,
    age: usize,
}

#[derive(serde::Serialize, serde::Deserialize)]
struct Users {
    users: Vec<User>,
}

fn bench_json_parse(c: &mut Criterion) {
    let users = Users {
        users: (0..JSON_N)
            .map(|i| User {
                id: i,
                name: format!("user_{}", i),
                email: format!("user{}@example.com", i),
                age: 20 + (i % 50),
            })
            .collect(),
    };
    let json_str = serde_json::to_string(&users).unwrap();

    let mut group = c.benchmark_group("JSON");
    group.throughput(Throughput::Bytes(json_str.len() as u64));

    group.bench_function("Parse", |b| {
        b.iter(|| {
            let _: Users = serde_json::from_str(&json_str).unwrap();
        })
    });

    group.finish();
}

// === 赛道 5: TOML ===

fn generate_toml_data() -> String {
    let mut s = String::with_capacity(TOML_N * 200);
    s.push_str("[server]\n");
    s.push_str("host = \"0.0.0.0\"\n");
    s.push_str("port = 8080\n");
    s.push_str("workers = 16\n");
    s.push_str("max_connections = 10000\n\n");
    s.push_str("[database]\n");
    s.push_str("driver = \"postgresql\"\n");
    s.push_str("host = \"db.example.com\"\n");
    s.push_str("port = 5432\n");
    s.push_str("pool_size = 20\n\n");

    for i in 0..TOML_N {
        write!(s, "[[services]]\n").unwrap();
        write!(s, "name = \"service_{}\"\n", i).unwrap();
        if i % 2 == 0 {
            s.push_str("enabled = true\n");
        } else {
            s.push_str("enabled = false\n");
        }
        write!(s, "weight = {}\n", i % 100).unwrap();
        write!(s, "timeout = {}\n", 1000.0 + (i as f64) * 10.0 + 0.5).unwrap();
        write!(s, "endpoint = \"https://api.example.com/v1/service_{}\"\n", i).unwrap();
        write!(s, "tags = [\"production\", \"region_{}\", \"tier_backend\"]\n\n", i % 5).unwrap();
    }
    s
}

#[derive(serde::Deserialize)]
struct TomlServer {
    host: String,
    port: u16,
    workers: u32,
    max_connections: u32,
}

#[derive(serde::Deserialize)]
struct TomlDatabase {
    driver: String,
    host: String,
    port: u16,
    pool_size: u32,
}

#[derive(serde::Deserialize)]
struct TomlService {
    name: String,
    enabled: bool,
    weight: u32,
    timeout: f64,
    endpoint: String,
    tags: Vec<String>,
}

#[derive(serde::Deserialize)]
struct TomlConfig {
    server: TomlServer,
    database: TomlDatabase,
    services: Vec<TomlService>,
}

fn bench_toml_parse(c: &mut Criterion) {
    let data = generate_toml_data();

    let mut group = c.benchmark_group("TOML");
    group.throughput(Throughput::Bytes(data.len() as u64));

    group.bench_function("Parse", |b| {
        b.iter(|| {
            let _: TomlConfig = toml::from_str(&data).unwrap();
        })
    });

    group.finish();
}

// === 赛道 6: Regex ===

fn generate_regex_log_lines() -> Vec<String> {
    (0..REGEX_N)
        .map(|i| {
            format!(
                "2026-06-29 14:3{}:2{}.{:03} [info] Request #{} completed in {}ms",
                i % 10,
                i % 6,
                100 + (i * 37) % 900,
                i,
                i % 500
            )
        })
        .collect()
}

fn bench_regex_match(c: &mut Criterion) {
    let re = regex::Regex::new(r"(\d{4}-\d{2}-\d{2}) (\d{2}:\d{2}:\d{2}\.\d{3}) \[(\w+)\] (.+)")
        .unwrap();
    let lines = generate_regex_log_lines();

    let mut group = c.benchmark_group("Regex");
    group.throughput(Throughput::Elements(REGEX_N as u64));

    group.bench_function("Match", |b| {
        b.iter(|| {
            let mut matched = 0usize;
            for line in &lines {
                if re.is_match(line) {
                    matched += 1;
                }
            }
            matched
        })
    });

    group.finish();
}

fn bench_regex_simple_match(c: &mut Criterion) {
    let re = regex::Regex::new(r"\d+").unwrap();
    let lines = generate_regex_log_lines();

    let mut group = c.benchmark_group("Regex");
    group.throughput(Throughput::Elements(REGEX_N as u64));

    group.bench_function("SimpleMatch", |b| {
        b.iter(|| {
            let mut matched = 0usize;
            for line in &lines {
                if re.is_match(line) {
                    matched += 1;
                }
            }
            matched
        })
    });

    group.finish();
}

fn bench_regex_findall(c: &mut Criterion) {
    let re = regex::Regex::new(r"(\d{4}-\d{2}-\d{2}) (\d{2}:\d{2}:\d{2}\.\d{3}) \[(\w+)\] (.+)")
        .unwrap();
    let lines = generate_regex_log_lines();

    let mut group = c.benchmark_group("Regex");
    group.throughput(Throughput::Elements(REGEX_N as u64));

    group.bench_function("FindAll", |b| {
        b.iter(|| {
            let mut total = 0usize;
            for line in &lines {
                let matches: Vec<_> = re.find_iter(line).collect();
                total += matches.len();
            }
            total
        })
    });

    group.finish();
}

criterion_group!(
    benches,
    bench_hashmap_insert,
    bench_hashmap_lookup,
    bench_hashmap_iterate,
    bench_sort_int32,
    bench_string_builder,
    bench_string_concat,
    bench_json_parse,
    bench_toml_parse,
    bench_regex_match,
    bench_regex_simple_match,
    bench_regex_findall,
);
criterion_main!(benches);
